packages/deployments/contracts/contracts/messaging/RootManager.sol

```diff
@@ -184,6 +185,32 @@ contract RootManager is ProposedOwnable, IRootManager, WatcherClient, DomainInde
    */
   function aggregate(uint32 _domain, bytes32 _inbound) external whenNotPaused onlyConnector(_domain) {
     uint128 lastIndex = pendingInboundRoots.enqueue(_inbound);
-    emit RootAggregated(_domain, _inbound, lastIndex);
+    emit RootReceived(_domain, _inbound, lastIndex);
+  }
+
+  /**
+   * @notice Dequeue verified inbound roots and insert them into the aggregator tree.
+   * @dev Will dequeue a fixed maximum amount of roots to prevent out of gas errors. As such, this
+   * method is public and separate from `propagate` so we can curtail an overloaded queue as needed.
+   * @dev Reverts if no verified inbound roots are found.
+   *
+   * @return bytes32 The new aggregate root.
+   * @return uint256 The updated count (number of leaves).
+   */
+  function dequeue() public whenNotPaused returns (bytes32, uint256) {
+    // Get all of the verified roots from the queue.
+    bytes32[] memory _verifiedInboundRoots = pendingInboundRoots.dequeueVerified(delayBlocks, DEQUEUE_MAX);
+
+    // Sanity check: there must be some verified roots to aggregate and send: otherwise we would be
+    // propagating a redundant aggregate root.
+    require(_verifiedInboundRoots.length != 0, "no verified roots");
```

Comment:
This can result in a DOS vector.
Suppose someone repeatedly calls `dequeue()` until the `queue` is empty (which can be done because its permissionless).

Then a relayer calls `propagate()`, which calls `dequeue()`.
As `dequeue()` reverts (because the queue is empty), `propagate()` will also revert.

Possible solution: just return when `_verifiedInboundRoots.length == 0`

Reply:
Oh excellent point! Yes, I think we can just return here when the queue is empty.

One issue this presents - propagate can be called even if no new roots have been added since the last propagate. I think I may need to add a flag to ensure we aren't calling a redundant propagate, since this check here:
```require(_verifiedInboundRoots.length != 0, "no verified roots");```
was previously the thing that prevented that from happening.